lib/vagrant/util/line_buffer.rb

```diff
@@ -0,0 +1,61 @@
+module Vagrant
+  module Util
+    class LineBuffer
+
+      # Maximum number of characters to buffer before sending
+      # to callback without detecting a new line
+      MAX_LINE_LENGTH = 5000.freeze
+
+      # Create a new line buffer. The registered block
+      # will be called when a new line is encountered on
+      # provided input, or the max line length is reached
+      def initialize(&callback)
+        raise ArgumentError,
+          "Expected callback but received none" if callback.nil?
+        @mu = Mutex.new
+        @callback = callback
+        @buffer = ""
+      end
+
+      # Add string data to output
+      #
+      # @param [String] str String of data to output
+      # @return [self]
+      def <<(str)
+        @mu.synchronize do
+          while i = str.index("\n")
+            @callback.call((@buffer + str[0, i+1]).rstrip)
+            @buffer.clear
+            str = str[i+1, str.length]
+          end
+
+          @buffer << str
+
+          if @buffer.length > MAX_LINE_LENGTH
+            @callback.call(@buffer.dup)
+            @buffer.clear
+          end
+        end
+        self
+      end
+
+      # Closes the buffer. Any remaining data that has
+      # been buffered will be given to the callback.
+      # Once closed the instance will no longer be usable.
+      #
+      # @return [self]
+      def close
+        @mu.synchronize do
+          # Send any remaining output on the buffer
+          @callback.call(@buffer.dup) if
+            !@buffer.empty?
```

Comment:
Can this be replaced with `unless @buffer.empty?` in the same line as the callback call?

Reply:
I pulled it up on the same line, but retained the `if !` instead of using `unless`. While there are instances of `unless` used in some of the Vagrant code, it's generally discouraged as `if` based conditions are easier to parse (for a person) the majority of the time, especially when there are compound statements.